tensorpipe/test/peer_group.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <string>
+#include <thread>
+
+#include <unistd.h>
+
+#include <gtest/gtest.h>
+
+class PeerGroup {
+ public:
+  static constexpr int kServer = 0;
+  static constexpr int kClient = 1;
+
+  virtual ~PeerGroup() = default;
+  virtual void sendPeer(int peerId, const std::string&) = 0;
+  virtual std::string recvPeer(int peerId) = 0;
+  virtual void spawn(std::function<void()>, std::function<void()>) = 0;
+
+  void donePeer(int peerId) {
+    sendPeer(1 - peerId, kDone);
+    std::unique_lock<std::mutex> lock(m_);
+    done_[peerId] = true;
+    cond_var_[peerId].notify_one();
+  }
+
+  void joinPeer(int peerId) {
+    while (recvPeer(peerId) != kDone)
+      ;
+    std::unique_lock<std::mutex> lock(m_);
+    while (!done_[peerId]) {
+      cond_var_[peerId].wait(lock);
+    }
+  }
+
+ private:
+  const std::string kDone = "done";
+  std::mutex m_;
+  bool done_[2] = {false, false};
+  std::condition_variable cond_var_[2];
+};
+
+class ThreadPeerGroup : public PeerGroup {
+ public:
+  void sendPeer(int peerId, const std::string& str) override {
+    q_[peerId].push(str);
+  }
+
+  std::string recvPeer(int peerId) override {
+    return q_[peerId].pop();
+  }
+
+  void spawn(std::function<void()> f1, std::function<void()> f2) override {
+    std::function<void()> fns[] = {std::move(f1), std::move(f2)};
+    std::vector<std::thread> ts;
+
+    for (auto& f : fns) {
+      ts.emplace_back(f);
+    }
+
+    for (auto& t : ts) {
+      t.join();
+    }
+  }
+
+ private:
+  tensorpipe::Queue<std::string> q_[2];
+};
+
+class ProcessPeerGroup : public PeerGroup {
+ public:
+  ProcessPeerGroup() {
+    for (int i = 0; i < 2; ++i) {
+      pipe(pipefd[i]);
+    }
+  }
+
+  ~ProcessPeerGroup() override {
+    for (int i = 0; i < 2; ++i) {
+      for (int j = 0; j < 2; ++j) {
+        close(pipefd[i][j]);
+      }
+    }
+  }
+
+  void sendPeer(int peerId, const std::string& str) override {
+    write(pipefd[peerId][1], str.c_str(), str.length());
+    write(pipefd[peerId][1], "\n", 1);
+  }
+
+  std::string recvPeer(int peerId) override {
+    std::string str;
+    char c;
+    while (read(pipefd[peerId][0], &c, 1) > 0) {
+      if (c == '\n') {
+        break;
+      }
+      str.push_back(c);
+    }
+
+    return str;
+  }
+
+  void spawn(std::function<void()> f1, std::function<void()> f2) override {
+    std::function<void()> fns[] = {std::move(f1), std::move(f2)};
+    pid_t pids[] = {-1, -1};
+
+    for (int i = 0; i < 2; ++i) {
+      pids[i] = fork();
```

Comment:
After we fork do we perhaps want to close some of the pipes? The way this currently works, is that all processes (parent, child0 and child1) all keep all the pipes open. If now child1 dies unexpectedly, the parent won't notice (because it is first waiting for child0 to finish, and it is still alive), and child0 may be blocked on a read from a pipe from child1. That read will never return because no one will ever write on that pipe. The read won't fail either, because both child0 and parent still have that pipe open and thus could potentially write to it. Thus we're in a deadlock.

A solution to the above problem is for the parent to immediately close all the pipes, and for each child to close one endpoint of each pipe. This way each endpoint is only open in exactly one process. That way, if that process dies unexpectedly, it will also in turn close one end of the pipe (no one else is keeping it alive) and thus anyone reading from the other end would correctly get an error.